File: reference/CodeExamples/dotNet/CSharpExamples/Code/pitfalls/updatedepth/UpdateDepthPitfall.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Linq;

namespace Db4oDoc.Code.Pitfalls.UpdateDepth
{
    public class UpdateDepthPitfall
    {
        public const String DatabaseFile = "database.db4o";

        public static void Main(string[] args)
        {
            CleanUp();
            PrepareDeepObjGraph();


            ToLowUpdateDeph();
            UpdateDepth();
        }

        private static void ToLowUpdateDeph()
        {
            // #example: Update doesn't work
            using (IObjectContainer container = Db4oEmbedded.OpenFile(DatabaseFile))
            {
                Person jodie = QueryForJodie(container);
                jodie.Add(new Person("Jamie"));
                // Remember that a collection is also a regular object
                // so with the default-update depth of one, only the changes
                // on the person-object are stored, but not the changes on
                // the friend-list.
                container.Store(jodie);
            }
            using (IObjectContainer container = Db4oEmbedded.OpenFile(DatabaseFile))
            {
                Person jodie = QueryForJodie(container);
                foreach (Person person in jodie.Friends)
                {
                    // the added friend is gone, because the update-depth is to low
                    Console.WriteLine("Friend=" + person.Name);
                }
            }
            // #end example
        }

        private static void UpdateDepth()
        {
            // #example: A higher update depth fixes the issue
            IEmbeddedConfiguration config = Db4oEmbedded.NewConfiguration();
            config.Common.UpdateDepth = 2;
            using (IObjectContainer container = Db4oEmbedded.OpenFile(config, DatabaseFile))
            {
                Person jodie = QueryForJodie(container);
                jodie.Add(new Person("Jamie"));
                // Remember that a collection is also a regular object
                // so with the default-update depth of one, only the changes
                // on the person-object are stored, but not the changes on
                // the friend-list.
                container.Store(jodie);
            }
            config = Db4oEmbedded.NewConfiguration();
            config.Common.UpdateDepth = 2;
            using (IObjectContainer container = Db4oEmbedded.OpenFile(config, DatabaseFile))
            {
                Person jodie = QueryForJodie(container);
                foreach (Person person in jodie.Friends)
                {
                    // the added friend is gone, because the update-depth is to low
                    Console.WriteLine("Friend=" + person.Name);
                }
            }
        }

        private static void CleanUp()
        {
            File.Delete(DatabaseFile);
        }

        private static Person QueryForJodie(IObjectContainer container)
        {
            return (from Person p in container
                    where p.Name == "Jodie"
                    select p).First();
        }

        private static void PrepareDeepObjGraph()
        {
            using (IObjectContainer container = Db4oEmbedded.OpenFile(DatabaseFile))
            {
                Person jodie = new Person("Jodie");

                jodie.Add(new Person("Joanna"));
                jodie.Add(new Person("Julia"));
                container.Store(jodie);
            }
        }
    }

    internal class Person
    {
        private IList<Person> friends = new List<Person>();

        private string name;

        internal Person(string name)
        {
            this.name = name;
        }


        public IList<Person> Friends
        {
            get { return friends; }
        }

        public string Name
        {
            get { return name; }
        }

        public void Add(Person item)
        {
            friends.Add(item);
        }
    }
}

File: reference/CodeExamples/dotNet/VisualBasicExamples/Code/pitfalls/updatedepth/UpdateDepthPitfall.vb
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports Db4objects.Db4o
Imports Db4objects.Db4o.Config
Imports Db4objects.Db4o.Linq

Namespace Db4oDoc.Code.Pitfalls.UpdateDepth
    Public Class UpdateDepthPitfall
        Public Const DatabaseFile As [String] = "database.db4o"

        Public Shared Sub Main(ByVal args As String())
            CleanUp()
            PrepareDeepObjGraph()


            ToLowUpdateDeph()
            UpdateDepth()
        End Sub

        Private Shared Sub ToLowUpdateDeph()
            ' #example: Update doesn't work
            Using container As IObjectContainer = Db4oEmbedded.OpenFile(DatabaseFile)
                Dim jodie As Person = QueryForJodie(container)
                jodie.Add(New Person("Jamie"))
                ' Remember that a collection is also a regular object
                ' so with the default-update depth of one, only the changes
                ' on the person-object are stored, but not the changes on
                ' the friend-list.
                container.Store(jodie)
            End Using
            Using container As IObjectContainer = Db4oEmbedded.OpenFile(DatabaseFile)
                Dim jodie As Person = QueryForJodie(container)
                For Each person As Person In jodie.Friends
                    ' the added friend is gone, because the update-depth is to low
                    Console.WriteLine("Friend=" & person.Name)
                Next
            End Using
            ' #end example
        End Sub

        Private Shared Sub UpdateDepth()
            ' #example: A higher update depth fixes the issue
            Dim config As IEmbeddedConfiguration = Db4oEmbedded.NewConfiguration()
            config.Common.UpdateDepth = 2
            Using container As IObjectContainer = Db4oEmbedded.OpenFile(config, DatabaseFile)
                Dim jodie As Person = QueryForJodie(container)
                jodie.Add(New Person("Jamie"))
                ' Remember that a collection is also a regular object
                ' so with the default-update depth of one, only the changes
                ' on the person-object are stored, but not the changes on
                ' the friend-list.
                container.Store(jodie)
            End Using
            config = Db4oEmbedded.NewConfiguration()
            config.Common.UpdateDepth = 2
            Using container As IObjectContainer = Db4oEmbedded.OpenFile(config, DatabaseFile)
                Dim jodie As Person = QueryForJodie(container)
                For Each person As Person In jodie.Friends
                    ' the added friend is gone, because the update-depth is to low
                    Console.WriteLine("Friend=" & person.Name)
                Next
            End Using
        End Sub

        Private Shared Sub CleanUp()
            File.Delete(DatabaseFile)
        End Sub

        Private Shared Function QueryForJodie(ByVal container As IObjectContainer) As Person
            Return (From p As Person In container Where p.Name = "Jodie").First()
        End Function

        Private Shared Sub PrepareDeepObjGraph()
            Using container As IObjectContainer = Db4oEmbedded.OpenFile(DatabaseFile)
                Dim jodie As New Person("Jodie")

                jodie.Add(New Person("Joanna"))
                jodie.Add(New Person("Julia"))
                container.Store(jodie)
            End Using
        End Sub
    End Class

    Friend Class Person
        Private m_friends As IList(Of Person) = New List(Of Person)()

        Private m_name As String

        Friend Sub New(ByVal name As String)
            Me.m_name = name
        End Sub


        Public ReadOnly Property Friends() As IList(Of Person)
            Get
                Return m_friends
            End Get
        End Property

        Public ReadOnly Property Name() As String
            Get
                Return m_name
            End Get
        End Property

        Public Sub Add(ByVal item As Person)
            m_friends.Add(item)
        End Sub
    End Class
End Namespace